### PR TITLE
Script to remove cache_filepath (and other) extras

### PR DESCRIPTION
**Pull request by morty**

### CODE REVIEW

File: ckanext/dgu/bin/delete_cache_filepath.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                                     .join(model.ResourceGroup)\
+                                     .join(model.Package)\
+                                     .filter(model.Package.state == 'active') \
+                                     .filter(model.Package.private == False) \
+                                     .filter(model.Resource.state == 'active')
+
+        for resource in all_resources:
+            if 'cache_filepath' in resource.extras:
+                stats.add('Removing cache_filepath', resource.id)
+                del resource.extras['cache_filepath']
```
Comment (morty) [inline ckanext/dgu/bin/delete_cache_filepath.py]: @davidread why doesn't this work like it does with package extras? Can't find any documentation on the right way to do it.

Comment (rossjones) [inline ckanext/dgu/bin/delete_cache_filepath.py]: Think you need to save the resource again as extras are just a field on the model (unlike other places where it is a separate model).

**Comment by davidread** [comment]
I tried your script and it seemed to delete the resource extra fine, so maybe you forgot to do `-w` or something.

I've extended it do delete a whole bunch of other legacy fields. Am running it now on prod2 and assuming we don't get any exceptions in the next couple of days we can merge and run it on live.

```
(ckan)co@prod2 /vagrant/src/ckanext-dgu (master) $ python ckanext/dgu/bin/delete_cache_filepath.py $CKAN_INI -w
85683 resources (active)
Resource Properties:
        Making property null: last_modified: 54066 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Making property null: size: 53879 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Making property null: mimetype: 53816 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Making property null: hash: 53793 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Making property null: cache_url: 52306 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Making property null: cache_last_updated: 52306 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Property has no value: cache_last_updated: 33377 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
        Property has no value: cache_url: 33377 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
        Property has no value: hash: 31890 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
        Property has no value: mimetype: 31867 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
        Property has no value: size: 31804 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
        Property has no value: last_modified: 31617 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
Resource Extras:
        No field to remove: content_length: 80898 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        No field to remove: content_type: 80898 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        No field to remove: openness_score: 79966 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        No field to remove: openness_score_failure_count: 79966 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        No field to remove: openness_score_reason: 79966 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        Removing: cache_filepath: 51104 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
        No field to remove: cache_filepath: 34579 [u'1761e957-fac2-4485-a371-8ce334a4a915', u'88bc00fb-cf29-477c-b272-8618bbe1a853', u'866f2d13-8a4a-4784-b0d1-6ed207a23cce', u'e2c1d06f-33c2-4038...
        Removing: openness_score: 5717 [u'9ddf7a95-1a63-4a94-bcc2-f73ada1f28d4', u'dea47feb-adbd-49ab-941c-e4e19e2fb6fb', u'263a5d26-d6c1-42e8-befe-7452636aef1f', u'15423a6d-5bfc-4d50-...
        Removing: openness_score_failure_count: 5717 [u'9ddf7a95-1a63-4a94-bcc2-f73ada1f28d4', u'dea47feb-adbd-49ab-941c-e4e19e2fb6fb', u'263a5d26-d6c1-42e8-befe-7452636aef1f', u'15423a6d-5bfc-4d50-...
        Removing: openness_score_reason: 5717 [u'9ddf7a95-1a63-4a94-bcc2-f73ada1f28d4', u'dea47feb-adbd-49ab-941c-e4e19e2fb6fb', u'263a5d26-d6c1-42e8-befe-7452636aef1f', u'15423a6d-5bfc-4d50-...
        Removing: content_length: 4785 [u'9ddf7a95-1a63-4a94-bcc2-f73ada1f28d4', u'dea47feb-adbd-49ab-941c-e4e19e2fb6fb', u'263a5d26-d6c1-42e8-befe-7452636aef1f', u'15423a6d-5bfc-4d50-...
        Removing: content_type: 4785 [u'9ddf7a95-1a63-4a94-bcc2-f73ada1f28d4', u'dea47feb-adbd-49ab-941c-e4e19e2fb6fb', u'263a5d26-d6c1-42e8-befe-7452636aef1f', u'15423a6d-5bfc-4d50-...
        Time taken (h:m:s): 0:00:16.129968
25283 datasets (active)
Dataset Properties:
        None
Dataset Extras:
        No field to remove: agency: 23306 [u'04x-25k', u'05p-25k-july-13', u'05p-25k-june-2013', u'05p-december-25k-expenditure', u'05p-over-25k-may', u'060-data-stats', u'1-1-5m-scale-g...
        No field to remove: department: 22978 [u'04x-25k', u'05p-25k-july-13', u'05p-25k-june-2013', u'05p-december-25k-expenditure', u'05p-over-25k-may', u'060-data-stats', u'1-1-5m-scale-g...
        No field to remove: openness_score_last_checked: 22782 [u'04x-25k', u'05p-25k-july-13', u'05p-25k-june-2013', u'05p-december-25k-expenditure', u'05p-over-25k-may', u'060-data-stats', u'1-1-5m-scale-g...
        No field to remove: openness_score: 22782 [u'04x-25k', u'05p-25k-july-13', u'05p-25k-june-2013', u'05p-december-25k-expenditure', u'05p-over-25k-may', u'060-data-stats', u'1-1-5m-scale-g...
        No field to remove: published_via: 20964 [u'04x-25k', u'05p-25k-july-13', u'05p-25k-june-2013', u'05p-december-25k-expenditure', u'05p-over-25k-may', u'060-data-stats', u'1-1-5m-scale-g...
        No field to remove: published_by: 20197 [u'04x-25k', u'05p-25k-july-13', u'05p-25k-june-2013', u'05p-december-25k-expenditure', u'05p-over-25k-may', u'060-data-stats', u'1-1-5m-scale-g...
        Removing: published_by: 5086 [u'13-butadiene-running-annual-mean-at-automatic-sites-comparison-with-health-objective-for-2003-u-2010', u'2008_injury_road_traffic_collisions_i...
        Removing: published_via: 4319 [u'2008_injury_road_traffic_collisions_in_northern_ireland', u'2050-pathways-call-for-evidence', u'a_generational_accounts_approach_to_long_term_...
        Removing: openness_score: 2501 [u'13-butadiene-running-annual-mean-at-automatic-sites-comparison-with-health-objective-for-2003-u-2010', u'2010-finance-survey-of-mid-cap-busine...
        Removing: openness_score_last_checked: 2501 [u'13-butadiene-running-annual-mean-at-automatic-sites-comparison-with-health-objective-for-2003-u-2010', u'2010-finance-survey-of-mid-cap-busine...
        Removing: department: 2305 [u'13-butadiene-running-annual-mean-at-automatic-sites-comparison-with-health-objective-for-2003-u-2010', u'2050-pathways-call-for-evidence', u'a...
        Removing: agency: 1977 [u'13-butadiene-running-annual-mean-at-automatic-sites-comparison-with-health-objective-for-2003-u-2010', u'2050-pathways-call-for-evidence', u'a...
        Time taken (h:m:s): 0:01:30.057715
```

**Comment by morty** [comment]
@davidread are you sure? On my VM it's still not removing them. Multiple runs show the same numbers:

```
(ckan)co@precise64:/src/ckanext-dgu$ python ckanext/dgu/bin/delete_cache_filepath.py /var/ckan/ckan.ini  -w 2>&1 | grep cache_filepath
	Removing: cache_filepath: 49761 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
	No field to remove: cache_filepath: 39178 [u'10e636e0-9d67-42ae-a952-c328d0858c14', u'2ef12c58-c789-4979-b119-63c4a048a5e1', u'e2c1d06f-33c2-4038-9466-20e0669bb0c6', u'16eff04b-e2cc-442e...
(ckan)co@precise64:/src/ckanext-dgu$ python ckanext/dgu/bin/delete_cache_filepath.py /var/ckan/ckan.ini  -w 2>&1 | grep cache_filepath
	Removing: cache_filepath: 49761 [u'8f034e7c-f137-4536-ba05-84e7a7975af2', u'476e8a45-b984-41a0-8dfc-6d982ae09c12', u'e8cfb8ab-5f1a-4418-8204-dd8f3a4bbe63', u'7427fa7d-f9e1-4ded...
	No field to remove: cache_filepath: 39178 [u'10e636e0-9d67-42ae-a952-c328d0858c14', u'2ef12c58-c789-4979-b119-63c4a048a5e1', u'e2c1d06f-33c2-4038-9466-20e0669bb0c6', u'16eff04b-e2cc-442e...
```

**Comment by rossjones** [comment]
This runs fine on my VM, it just takes a long time to write.  Is it worth trying this on prod2?

During the run through, that process ended up using 3.6GB of RAM, which isn't an issue because it's a one-off task, but worth bearing in mind if you have a low ram VM.

**Comment by davidread** [comment]
When I tried it on prod2 before it didn't work, so investigation is needed on there (hint)

**Comment by rossjones** [comment]
It appears that deleting the resource extras dict does not cause it to be marked for update in the session - I had to deepcopy the extras, and modify them.  Can someone give this a try (and a review) just to check my sanity please?

**Comment by morty** [comment]
This seems to work well on my VM.

Took about an hour to finish.

**Comment by davidread** [comment]
Great! Makes sense to run on prod2 then prod3?

**Pull request merged**